pandas/core/indexes/datetimes.py

```diff
@@ -252,19 +253,55 @@ def __new__(cls, data=None,
         # if dtype has an embedded tz, capture it
         tz = dtl.validate_tz_from_dtype(dtype, tz)
 
-        if not isinstance(data, (np.ndarray, Index, ABCSeries, DatetimeArray)):
-            # other iterable of some kind
-            if not isinstance(data, (list, tuple)):
+        if not hasattr(data, "dtype"):
+            # e.g. list, tuple
+            if np.ndim(data) == 0:
+                # i.e. generator
                 data = list(data)
-            data = np.asarray(data, dtype='O')
+            data = np.asarray(data)
+            copy = False
         elif isinstance(data, ABCSeries):
             data = data._values
 
-        # data must be Index or np.ndarray here
-        if not (is_datetime64_dtype(data) or is_datetimetz(data) or
-                is_integer_dtype(data) or lib.infer_dtype(data) == 'integer'):
-            data = tools.to_datetime(data, dayfirst=dayfirst,
-                                     yearfirst=yearfirst)
+        # By this point we are assured to have either a numpy array or Index
+
+        if is_float_dtype(data):
+            # Note: we must cast to datetime64[ns] here in order to treat these
+            #  as wall-times instead of UTC timestamps.
+            data = data.astype(_NS_DTYPE)
+            copy = False
+            # TODO: Why do we treat this differently from integer dtypes?
+
+        elif is_timedelta64_dtype(data):
+            warnings.warn("Passing timedelta64-dtype data to {cls} is "
+                          "deprecated, will raise a TypeError in a future "
+                          "version".format(cls=cls.__name__),
+                          FutureWarning, stacklevel=2)
+            data = data.view(_NS_DTYPE)
+
+        elif is_period_dtype(data):
+            # Note: without explicitly raising here, PeriondIndex
+            #  test_setops.test_join_does_not_recur fails
+            raise TypeError("Passing PeriodDtype data to {cls} is invalid.  "
+                            "Use `data.to_timestamp()` instead"
+                            .format(cls=cls.__name__))
+
+        elif is_extension_type(data) and not is_datetime64tz_dtype(data):
+            # Includes categorical
+            # TODO: We have no tests for these
+            data = np.array(data, dtype=np.object_)
+            copy = False
+
+        if is_object_dtype(data) or is_string_dtype(data):
+            # TODO: We do not have tests specific to string-dtypes,
+            #  also complex or categorical or other extension
+            copy = False
+            if lib.infer_dtype(data) == 'integer':
```

Comment:
IIRC we had to do this because we would cast a list of integers to a numpy array with dtype object; therefore any dtype inspection would be object instead of integer.

I see that you've modified that casting above so we may not need this case anymore?

Reply:
This is still needed.  There is a test that specifically passes `np.array([0], dtype=object)` which would fail if we didn't special-case the inferred-dtype.

Reply:
yeah this is ok